contracts/TwabRewards.sol

```diff
@@ -0,0 +1,379 @@
+    mapping(uint256 => mapping(uint256 => mapping(address => uint256))) internal _claimedRewards;
+
+    /* ============ Events ============ */
+
+    /**
+        @notice Emitted when contract has been deployed.
+        @param owner Contract owner address
+   */
+    event Deployed(address owner);
+
+    /**
+        @notice Emmited when a promotion is created.
+        @param promotion Pomotion settings
+    */
+    event PromotionCreated(Promotion promotion);
+
+    /* ============ Constructor ============ */
+
+    /**
+        @notice Deploy TwabRewards contract.
+        @param _owner Contract owner address
+    */
+    constructor(address _owner) Ownable(_owner) {
+        emit Deployed(_owner);
+    }
+
+    /* ============ Modifiers ============ */
+
+    /// @dev Ensure that the caller is the creator of the currently active promotion.
+    modifier onlyPromotionCreator() {
+        require(msg.sender == _getCurrentPromotion().creator, "TwabRewards/only-promotion-creator");
+        _;
+    }
+
+    /* ============ External Functions ============ */
+
+    /// @inheritdoc ITwabRewards
+    function createPromotion(address _ticket, PromotionParameters calldata _promotionParameters)
+        external
+        override
+        returns (bool)
+    {
+        require(_isPromotionActive() == false, "TwabRewards/promotion-already-active");
+
+        _requireTicket(_ticket);
+
+        uint256 _nextPromotionId = _currentPromotionId + 1;
+        _currentPromotionId = _nextPromotionId;
+
+        address _token = _promotionParameters.token;
+        uint256 _tokensPerEpoch = _promotionParameters.tokensPerEpoch;
+        uint256 _numberOfEpochs = _promotionParameters.numberOfEpochs;
+
+        IERC20(_token).safeTransfer(address(this), _tokensPerEpoch * _numberOfEpochs);
+
+        Promotion memory _nextPromotion = Promotion(
+            uint32(_nextPromotionId),
+            false,
+            msg.sender,
+            _ticket,
+            _token,
+            _tokensPerEpoch,
+            _promotionParameters.startTimestamp,
+            _promotionParameters.epochDuration,
+            _numberOfEpochs
+        );
+
+        _promotions[_nextPromotionId] = _nextPromotion;
+
+        emit PromotionCreated(_nextPromotion);
+
+        return true;
+    }
+
+    /// @inheritdoc ITwabRewards
+    function cancelPromotion(address _to) external override onlyPromotionCreator returns (bool) {
+        require(_isPromotionActive() == true, "TwabRewards/no-active-promotion");
+        require(_to != address(0), "TwabRewards/recipient-not-zero-address");
+
+        Promotion memory _currentPromotion = _getCurrentPromotion();
+
+        IERC20 _token = IERC20(_currentPromotion.token);
+        uint256 _remainingRewards = _getRemainingRewards(_token);
+
+        if (_remainingRewards > 0) {
+            _token.safeTransfer(_to, _remainingRewards);
+        }
+
+        _promotions[_currentPromotion.id].cancelled = true;
```

Comment:
Here you can save a bunch of gas by simply deleting the promotion

Reply:
So we don't want to be able to track past promotions that got cancelled?

Reply:
Done: https://github.com/pooltogether/v4-periphery/pull/8/commits/03349ffa05d8d3d3e37ce5fb54a49631caf7d753